cardinal/cardinal_test.go

```diff
@@ -40,6 +44,635 @@ type Rawbodytx struct {
+			return cardinal.UpdateComponent[CounterComponent](
+				wCtx, id, func(c *CounterComponent) *CounterComponent {
+					c.Count++
+					return c
+				},
+			)
+		},
+	)
+	assert.NilError(t, err)
+	tf.StartWorld()
+	id, err := cardinal.Create(wCtx, CounterComponent{})
+	assert.NilError(t, err)
+
+	for i := 0; i < 10; i++ {
+		tf.DoTick()
+	}
+
+	c, err := cardinal.GetComponent[CounterComponent](wCtx, id)
+	assert.NilError(t, err)
+	assert.Equal(t, 10, c.Count)
+}
+
+type ScoreComponent struct {
+	Score int
+}
+
+func (ScoreComponent) Name() string {
+	return "score"
+}
+
+type ModifyScoreMsg struct {
+	PlayerID types.EntityID
+	Amount   int
+}
+
+type EmptyMsgResult struct{}
+
+func TestTransactionAreAppliedToSomeEntities(t *testing.T) {
+	tf := testutils.NewTestFixture(t, nil)
+	world := tf.World
+	assert.NilError(t, cardinal.RegisterComponent[ScoreComponent](world))
+
+	assert.NilError(t, cardinal.RegisterMessage[*ModifyScoreMsg, *EmptyMsgResult](world, "modify_score"))
+
+	err := cardinal.RegisterSystems(
+		world,
+		func(wCtx engine.Context) error {
+			modifyScoreMsg, err := cardinal.GetMessage[*ModifyScoreMsg, *EmptyMsgResult](wCtx)
+			assert.NilError(t, err)
+			modifyScores := modifyScoreMsg.In(wCtx)
+			for _, msData := range modifyScores {
+				ms := msData.Msg
+				err := cardinal.UpdateComponent[ScoreComponent](
+					wCtx, ms.PlayerID, func(s *ScoreComponent) *ScoreComponent {
+						s.Score += ms.Amount
+						return s
+					},
+				)
+				assert.Check(t, err == nil)
+			}
+			return nil
+		},
+	)
+	assert.NilError(t, err)
+	tf.StartWorld()
+
+	wCtx := cardinal.NewWorldContext(world)
+	ids, err := cardinal.CreateMany(wCtx, 100, ScoreComponent{})
+	assert.NilError(t, err)
+	// Entities at index 5, 10 and 50 will be updated with some values
+	modifyScoreMsg, err := cardinal.GetMessage[*ModifyScoreMsg, *EmptyMsgResult](wCtx)
+	assert.NilError(t, err)
+	tf.AddTransaction(
+		modifyScoreMsg.ID(), &ModifyScoreMsg{
+			PlayerID: ids[5],
+			Amount:   105,
+		},
+	)
+	tf.AddTransaction(
+		modifyScoreMsg.ID(), &ModifyScoreMsg{
+			PlayerID: ids[10],
+			Amount:   110,
+		},
+	)
+	tf.AddTransaction(
+		modifyScoreMsg.ID(), &ModifyScoreMsg{
+			PlayerID: ids[50],
+			Amount:   150,
+		},
+	)
+
+	tf.DoTick()
+
+	for i, id := range ids {
+		wantScore := 0
+		switch i {
+		case 5:
+			wantScore = 105
+		case 10:
+			wantScore = 110
+		case 50:
+			wantScore = 150
+		}
+		s, err := cardinal.GetComponent[ScoreComponent](wCtx, id)
+		assert.NilError(t, err)
+		assert.Equal(t, wantScore, s.Score)
+	}
+}
+
+// TestAddToPoolDuringTickDoesNotTimeout verifies that we can add a transaction to the transaction
+// pool during a game tick, and the call does not block.
+func TestAddToPoolDuringTickDoesNotTimeout(t *testing.T) {
+	tf := testutils.NewTestFixture(t, nil)
+	world := tf.World
+
+	assert.NilError(t, cardinal.RegisterMessage[*ModifyScoreMsg, *EmptyMsgResult](world, "modify_Score"))
+
+	inSystemCh := make(chan struct{})
```

Comment:
This fixes the deadlock:

```
	inSystemCh := make(chan struct{})
	// Push a value to inSystemCh to verify we're in the middle of a tick. Close inSystemCh at the end of the test
	// to make sure the System exits so that post-test cleanup can take place.
	defer func() { close(inSystemCh) }()
	err := cardinal.RegisterSystems(
		world,
		func(engine.Context) error {
			<-inSystemCh
			<-inSystemCh
			return nil
		},
	)
```

Reply:
Still blocks. I say let's fix this in another ticket. It detracts from the current issue I'm dealing with here. I think the refactor of the "ticks" was a bit of detraction but I was fine with it because it's small but it looks like it's ballooning slowly into something a bit bigger.